src/pure_aworset.erl

```diff
@@ -0,0 +1,257 @@
+%%
+%% Copyright (c) 2015-2016 Christopher Meiklejohn.  All Rights Reserved.
+%% Copyright (c) 2007-2012 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc Pure AWORSet CRDT: pure op-based add-wins observed-remove set
+%%
+%% @reference Carlos Baquero, Paulo Sérgio Almeida, and Ali Shoker
+%%      Making Operation-based CRDTs Operation-based (2014)
+%%      [http://haslab.uminho.pt/ashoker/files/opbaseddais14.pdf]
+
+-module(pure_aworset).
+-author("Georges Younes <georges.r.younes@gmail.com>").
+
+-behaviour(type).
+
+-define(TYPE, ?MODULE).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([new/0, new/1]).
+-export([mutate/3, query/1, equal/2]).
+
+-export_type([pure_aworset/0, pure_aworset_op/0]).
+
+-opaque pure_aworset() :: {?TYPE, payload()}.
+-type payload() :: {pure_type:polog(), ordsets:set()}.
+-type pure_aworset_op() :: {add, pure_type:element()} | {rmv, pure_type:element()}.
+
+%% @doc Create a new, empty `pure_aworset()'
+-spec new() -> pure_aworset().
+new() ->
+    {?TYPE, {orddict:new(), ordsets:new()}}.
+
+%% @doc Create a new, empty `pure_aworset()'
+-spec new([term()]) -> pure_aworset().
+new([]) ->
+    new().
+
+%% @doc Check causal order of 2 version vector.
+-spec happened_before(pure_type:id(), pure_type:id()) -> boolean().
+happened_before(VV1, VV2) ->
+    orddict:fold(
+        fun(Key, Value1, Acc) ->
+            case orddict:find(Key, VV2) of
+                {ok, Value2} ->
+                    Acc andalso Value1 =< Value2;
+                error ->
+                    Acc andalso false
+            end
+        end,
+        true,
+        VV1
+     ) andalso not (VV1 == VV2).
```

Comment:
No, this won't work.  You need to use the orddict equality function, because these data structures, since they are tree-based, are not guaranteed to be equivalent in structure.